activerecord/lib/active_record/core.rb

```diff
@@ -201,6 +201,15 @@ def self.current_preventing_writes
         false
       end
 
+      def self.preventing_writes?(class_name)
+        connected_to_stack.reverse_each do |hash|
+          return hash[:prevent_writes] if !hash[:prevent_writes].nil? && hash[:klasses].include?(Base)
+          return hash[:prevent_writes] if !hash[:prevent_writes].nil? && hash[:klasses].any? { |klass| klass.name == class_name }
```

Comment:
We'd need to do the same check in `current_preventing_writes`, `current_role` and `current_shard`

Reply:
Are you sure about that? The essential part of the bug described in #54343 is that `current_preventing_writes` was being called on the class cached in `PoolConfig`. 

There were no other methods being called on that class, which made it possible to replace it with a modified `StringConnectionName` instance. I believe the existing `current_*` methods in `core.rb` are fine because the classes being stored in the `connected_to_stack` are ephemeral (per-request) and not cached.